dali/operators/reader/loader/webdataset/tar_utils.h

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_READER_LOADER_TAR_UTILS_H_
+#define DALI_OPERATORS_READER_LOADER_TAR_UTILS_H_
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+#include "dali/core/common.h"
+#include "dali/util/file.h"
+
+namespace dali {
+namespace detail {
+/**
+ * @brief Used to access .tar archives
+ *
+ * The class is used to access tar archives through the FileStream of the user's choice.
+ * There are 2 cursors that it keeps track of:
+ *   - archive cursor - Keeps track of which file the archive is currently at.
+ *   - file cursor - Keeps track of which byte of the file is currently going to be read. Gets reset
+ *                   after each advancement of the archive cursor.
+ * There are also two associated sets of methods that the user may use:
+ *   - Files iteration methods:
+ *     - @ref NextFile - Advanced the archive cursor to the next file. Returns whether it
+ *                       has reached the end of archive.
+ *     - @ref EndOfArchive - Whether it has reached the end of archive.
+ *   - File contents access methods:
+ *     - @ref GetFileName - Returns the name of the file the archive cursor currently points at.
+ *     - @ref GetFileSize - Returns the size of the archive cursor currently points at (in bytes).
+ *     - @ref ReadFile - Reads the contents of the file and returns them. In the case of success
+ *                       places the file cursor at the end of file. In the other case keeps the file
+ *                       cursor intact.
+ *     - @ref Read - Reads the given number of bytes into a given buffer, assuming that said buffer
+ *                   is big enough to perform this operation. Returns the number of bytes actually
+ *                   read.
+ *     - @ref EndOfFile - Returns whether the file cursor is at the end of file.
+ */
```

Comment:
As methods are documented below I think that repeating it here does not add much value and makes the documentation harder to maintain. Maybe it should be removed?